# SCOUTS - Single Cell Outlier Selector
<p align="middle">
<img src="doc/_static/icons/scouts_icon_small.png" alt="SCOUTS icon">
</p>

**SCOUTS** SCOUTS is a tool that quickly finds **outliers in your single-cell data**, generating information about your cell population organized by markers. It is designed to analyse **mass cytometry (CyToF)** and **single-cell RNA sequencing** data (scRNASeq).

# Why SCOUTS?
Many single-cell analytical pipelines require some level of programming knowledge in order to be used. While some great tools for languages like R, Python and Julia have been developed, the entry-level barrier of programming is still intimidating for many scientists starting on the field of singe-cell analysis. With this in mind, we developed SCOUTS to simplify this process. Through a desktop application, the user is able to choose the parameters for the outlier selection, and leave the hard work of programmatically subsetting the data to SCOUTS.

As a showcase of how to interpret and explore the data generated by SCOUTS, we developed **SCOUTS-violins**, a secondary desktop application which displays the outlier populations identified by SCOUTS as violin plots.

# Getting Started
## Installation
SCOUTS is available as a:

* Python package from PyPI - install with `pip`
* Conda package install with `conda`
* GitHub repository - download/clone the repository
* binary release (experimental)

For any installation option (other than the binary release), SCOUTS requires **Python >= 3.6** to installed in your system. To check this, open a terminal/cmd and type:

```
$ python3 --version
```
If the output is something like `Python 3.6`, you're golden! Otherwise you may need to [download or upgrade Python](https://www.python.org/).

For PyPI/pip or GitHub installations, we recommend using a [virtual environment](https://docs.python.org/3/tutorial/venv.html), instead of installing SCOUTS into your global/system-wide Python interpreter.

### Install from PyPI using `pip`
To install SCOUTS, simply type in a terminal:

```
$ pip install scouts
```

then run SCOUTS by entering `scouts` in your terminal.

#### Optional: SCOUTS-violins
If you also want to install SCOUTS-violins, type:

```
$ pip install scouts[violins]
``` 

then run SCOUTS-violins by entering `scouts-violins` in your terminal.

### Install using `conda`
Coming soon!

### Install from the GitHub repository
Download this repository into your local machine. Alternatively, clone it with `git`:

```
$ git clone https://github.com/jfaccioni/scouts
```

Enter the repository's directory:

```
$ cd scouts
```

Make sure your Python interpreter (version >= 3.6) has the following packages installed:
* [numpy](http://www.numpy.org/)
* [pandas](https://pandas.pydata.org/)
* [pyside2](https://wiki.qt.io/Qt_for_Python)
* [openpyxl](https://openpyxl.readthedocs.io/en/stable/)
* [xlrd](https://xlrd.readthedocs.io/en/latest/)

Run SCOUTS by typing:

```
$ python scouts.py
```

#### Optional: SCOUTS-violins
If you also want to install SCOUTS-violins, make sure your Python interpreter (version >= 3.6) has the following additional packages installed:
* [matplotlib](https://matplotlib.org/)
* [seaborn](https://seaborn.pydata.org/)

Run `scouts-violins` by typing:

```
$ python scouts-violins.py
```

#### Using `pipenv` with the GitHub repository
For `pipenv` users, we included a `Pipfile` for convenience. Simply type `pipenv install` from the repository's directory to install SCOUTS into a virtual environment, along with the necessary third-party dependencies. This covers the installation of both SCOUTS and SCOUTS-violins.

### Download binary release and run as an executable
Download the binary release for your OS:

* [Windows](http://www.ufrgs.br/labsinal/scouts/scouts.exe)
* MacOS (coming soon!)
* [Linux](http://www.ufrgs.br/labsinal/scouts/scouts-linux)

If you choose this option, please be aware that:
* SCOUTS is a Python package. The executable has a moderately large size (~150 mb), since it has to bundle the whole Python interpreter along with it.
* this is an **experimental release of SCOUTS**, and as such it is not expected to support all OS configurations. If you run into any problems, please choose another installation method.

## Basic usage
The `scouts` package include two user interfaces:
* **SCOUTS**: used to select outliers in a population of single-cells.
* **SCOUTS-violins** (optional): used to visually inspect outliers selected by SCOUTS.

### Using SCOUTS
This is a basic rundown of the interface when you start SCOUTS: 

<p align="middle">
<img src="doc/\_static/gui/scouts_main_page_annotations_small.png" alt="SCOUTS main page"/>
</p>

1) Choose your input data
2) Choose your sample names - [explanation here](https://scouts.readthedocs.io/en/master/work.html)
3) Choose whether to gate samples and analyse other outlier populations - [explanation here](https://scouts.readthedocs.io/en/master/work.html)
4) Choose how to perform outlier selection - [explanation here](https://scouts.readthedocs.io/en/master/work.html)
5) Choose how and where to save output
6) Run the program

### Using SCOUTS-violins
This is a basic rundown of the interface when you start SCOUTS-violins: 

<p align="middle">
<img src="doc/\_static/gui/scouts_violins_main_page_annotations_small.png" alt="SCOUTS-violins main page"/>
</p>

1) Choose your input data (same input file used for SCOUTS + SCOUTS output folder)
2) Choose your sample names - [explanation here](https://scouts.readthedocs.io/en/master/work.html)
3) Choose plot parameters
4) Plot your data

# Documentation
[Read the full documentation here](https://scouts.readthedocs.io/en/master/).

# FAQ
[Frequently asked questions can be found here](https://scouts.readthedocs.io/en/master/faq.html).

# Citations
If you use SCOUTS, please [cite us](http://www.ufrgs.br/labsinal/).

# Authors
**Juliano Faccioni** - Programming and GUI development
* [GitHub](https://github.com/jfaccioni)
* [LinkedIn](https://www.linkedin.com/in/juliano-faccioni-9b2133167)

**Giovana Onzi** - Concept and testing
* [LinkedIn](https://www.linkedin.com/in/giovana-onzi-ba222895/)

# License
This project is licensed under the [MIT License](LICENSE.md).

# Acknowledgements
**Scientific counselling**
* Prof. Dr. Guido Lenz

**Funding**
* CAPES/CNPq
* NIH
